accord-core/src/main/java/accord/messages/SafeCallback.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package accord.messages;
+
+import java.util.Objects;
+import java.util.concurrent.Executor;
+
+import accord.api.Agent;
+import accord.coordinate.Timeout;
+import accord.local.Node;
+
+public class SafeCallback<T extends Reply>
+{
+    private final Executor executor;
+    private final Agent agent;
```

Comment:
Maybe we can extract a common parent interface of `CommandStore` with an `agent` method?

Reply:
Do you mean the following type setup?

```
Executor - java's
AsyncExecutor extends Executor - adds AsyncChain support
AgentExecutor extends AsyncExecutor - adds "Agent agent();"
CommandStore extends AsyncExecutor
```

Looking at the usage, since the sink API uses `Executor` then the sink needs access to the `Agent`, so is this idea to move away from `Executor` in favor of `AgentExecutor` which has a `agent`?

Reply:
Something like this, though I don't know that we need to implement the `Executor` interface at all, we might prefer a different method to `execute`.

Reply:
pushed change... I left the dependency of `Executor` alone as that is used directly in `SafeCallback`... we don't "need" the `async` methods, but `AgentExecutor` depending on `AsyncExecutor` I felt was still best...